flexget/utils/serialization.py

```diff
@@ -0,0 +1,128 @@
+import datetime
+from abc import ABC, abstractmethod
+
+from flexget.utils import json
+
+DATE_FMT = '%Y-%m-%d'
+ISO8601_FMT = '%Y-%m-%dT%H:%M:%SZ'
+
+
+def serialize(value):
+    """
+    Convert an object to JSON serializable format.
+    """
+    if isinstance(value, Serializable):
+        return value.serialize()
+    if isinstance(value, (list, tuple)):
+        return [serialize(v) for v in value]
+    if isinstance(value, dict):
+        return {k: serialize(v) for k, v in value.items()}
+    if isinstance(value, datetime.datetime):
+        return DateTimeSerializer.serialize(value)
+    if isinstance(value, datetime.date):
+        return DateSerializer.serialize(value)
+    if isinstance(value, (str, int, float, type(None))):
+        return value
+    raise TypeError(f'`{value!r}` of type {type(value)!r} is not serializable')
+
+
+def deserialize(value):
+    if isinstance(value, dict):
+        if all(key in value for key in ('serializer', 'version', 'value')):
+            return _registry()[value['serializer']].deserialize(value)
+        return {k: deserialize(v) for k, v in value.items()}
+    if isinstance(value, list):
+        return [deserialize(v) for v in value]
+    return value
+
+
+class Serializable(ABC):
+    @abstractmethod
+    def _serialize(self):
```

Comment:
I don't think abstractmethods are generally prefixed with _, which heavily implies internal functionality not to be messed with. More confusingly there is public `serialize` method which is NOT overridden.

Very confusing situation with `serialize` and `_serialize`.

Reply:
How are you thinking it should work? The idea was that `serialize` is the public method that can be called on any instances of a serializable object, and that `_serialize` is the private implementation, not called external to the class. I based it on how [marshmallow works](https://github.com/marshmallow-code/marshmallow/blob/dev/src/marshmallow/fields.py#L367)